pkg/proxy/iptables/proxier.go

```diff
@@ -611,30 +606,22 @@ func buildEndpointsMap(allEndpoints []api.Endpoints, hostname string, oldEndpoin
 						isLocalEndpoint = *addr.NodeName == hostname
 						isLocalEndpoint = featuregate.DefaultFeatureGate.ExternalTrafficLocalOnly() && isLocalEndpoint
 					}
-					hostPortObject := hostPortInfo{
-						host:          addr.IP,
-						port:          int(port.Port),
-						localEndpoint: isLocalEndpoint,
-					}
-					portsToEndpoints[port.Name] = append(portsToEndpoints[port.Name], hostPortObject)
+					addEndpointToMap(portsToEndpoints, addr.IP, port, isLocalEndpoint)
 					addHcUpdate(hcUpdates, namespacedName, isLocalEndpoint)
 				}
 			}
 		}
-		for portname, endpoints := range portsToEndpoints {
+		for portname, newEndpoints := range portsToEndpoints {
 			svcPort := proxy.ServicePortName{NamespacedName: namespacedName, Port: portname}
 			curEndpoints := newEndpointsMap[svcPort]
-			newEndpoints := flattenValidEndpoints(endpoints)
-			// Flatten the list of current endpoint infos to just a list of ips as strings
-			curEndpointIPs := flattenEndpointsInfo(curEndpoints)
-			if len(curEndpointIPs) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpointIPs), newEndpoints) {
-				removedEndpoints := getRemovedEndpoints(curEndpointIPs, newEndpoints)
+			sort.Sort(newEndpoints)
+			if !reflect.DeepEqual(curEndpoints, newEndpoints) || len(newEndpoints) == 0 {
```

Comment:
@danwinship you're right, it can be removed if we move the `newEndpointsMap[svcPort] = newEndpoints` out of the if{} block.